### ISSUE TITLE
Urukul v1.0 detailed testing

### ISSUE THREAD

**Issue by hartytp** [sinara-hw/sinara#354]
- [x] Channel-to-channel cross-talk: adjacent channels, with agressor, victim switches opened and switches closed and with attenuators at max/min
- [x] Output phase noise/amplitude noise (0.1 Hz - 10 MHz)
- [x] OIP3, OIP2 vs frequency for min attenuator setting and medium/large amplitude (ASF)
- [x] RF switch settling behavior for on edge at max power (worst case)
- [x] Attenuator glitches
- [x] Nyquist attenuation, clock leakage
- [ ] RF switch settling behavior for off edge at max power (worst case)
- [ ] RF switch phase chirp
- [ ] Temperature coefficient of power and phase
- [ ] Are we happy running the IC11 and running the LVDS interface directly from the 3V6 rail? (Try this and check performance)
- [ ] Check performance of clock buffers when running in LVDS mode from the 1V8A rail
- [x] Optimize PLL loop filters (@weidazhang can you do this for the AD9910 boards, please?)
- [ ] Board-to-board crosstalk (in proximity to Kasli, DIO pulsing 5 V into 50 R, Urukul)

Use some fixed reference frequency for all tests, such as 80MHz.

**Comment by hartytp** [comment]
@gkasprow Can you ship us all but 1 of the AD9910 Urukul prototypes you've ordered as soon as basic functionality/"no smoke" testing is complete (i.e. don't complete the above detailed testing first!). 

Thanks!

**Comment by hartytp** [comment]
@gkasprow @jordens Feel free to add more tests to the above list.

**Comment by hartytp** [comment]
@gkasprow When you ship the AD9910 DDSs that we paid for, please can you ship one to @jordens?

**Comment by jordens** [comment]
Updated it a bit. Please ship right after smoke test:

- [ ] power supplies good
- [ ] dds pll locks
- [ ] dds outputs frequency
- [ ] attenuators can be set
- [ ] rf switches work

We'll be happy to help with characterization.

**Comment by gkasprow** [comment]
@hartytp OK, 2 Zotinos + BNC adapters were shipped on Friday. Once I receive DDS, I will test them ASAP.

**Comment by hartytp** [comment]
Thanks!

**Comment by gkasprow** [comment]
@hartytp There are no panels for DAC BNC adapters attached - I forgotten about them. Will ship them next time with some other stuff.

**Comment by hartytp** [comment]
@gkasprow No problem. That can wait.

If there is any time before Urukul arrives, it'd be great to get a few of the analog test for Zotino done.

**Comment by gkasprow** [comment]
@hartytp Sure, once I repair my bard, will continue with measurements.

**Comment by hartytp** [comment]
Great, thanks @gkasprow!

**Comment by jordens** [comment]
@hartytp could you remember to do a bit of triage on the issues you file (i.e. assignee, labels, milestone)?

**Comment by hartytp** [comment]
sure

**Comment by hartytp** [comment]
@gkasprow For the current source biasing, my inclination is to go for Daniel's suggestion of the LT3092. They're pretty cheap ($2 in quantity) and seem like a better/easier solution than building our own current source.

@jordens @gkasprow if you're oaky with this approach, then we should buy an eval board at some point and test this out with the v1.0 boards we have. I'm happy to do this testing if that helps -- I probably won't do this until after Christmas, but I don't think we're in a rush for it.

**Comment by jordens** [comment]
@a-shafir and I will take care of the first round of noise, power, RF switch testing.

**Comment by hartytp** [comment]
Cool.

Out of curiosity, do you have a proper phase noise/amplitude noise meter? Or, will you just stick this on a spectrum analyser?

**Comment by gkasprow** [comment]
@jordens I have them all in my lab but was busy with Sayma Ethernet debugging. You can follow my fight here #854

**Comment by gkasprow** [comment]
So the next thing is 9912 testing.

**Comment by jordens** [comment]
I am following that and involved.
Thanks for the AD9912 update. Once you have smoke-tested them like the AD9910 stuff, you can send two of them to me.

**Comment by a-shafir** [comment]
@gkasprow what was actually flashed into the CPLD on the 9910 board we have in Berlin?

I see 2 options
https://github.com/m-labs/sinara/tree/master/ARTIQ_ALTIUM/EEMs/Urukul/CPLD_code
or
https://github.com/quartiq/urukul

I am testing h/w but can't generate any output on the IO_SESET (P129) and some pins around.

**Comment by a-shafir** [comment]
@gkasprow how you tested 3U DDS 9910?
Have you got DDS ic working?
Do you have the code that sent the spi commands, what actually sent to DDS/atten etc?

**Comment by gkasprow** [comment]
@a-shafir I burned Roberts Verilog code. The only change I did was assignment of test points. Now we have:
```
assign tp = (cfg_data_rst | ((~en_9910) & tstriple7_i));
assign tp_1 = dds_cs_n;
assign tp_2 = dds_sck;
assign tp_3 = dds_sdi;
assign tp_4 = dds_io_update;

```

**Comment by gkasprow** [comment]
I connected 9910 devkit via DIO-Tester module to make sure that Robert's code works. In this way I could communicate with it using LVTTL signals

**Comment by gkasprow** [comment]
I also modified the verilog code so all 4 channels were active simultaneously. In this way I tested if PLLs are locked. Here is the code:
[Urukul_CPLD_all_ch.zip](https://github.com/m-labs/sinara/files/1549674/Urukul_CPLD_all_ch.zip)

**Comment by gkasprow** [comment]
To do general purpose communication over SPI I use little Artix 7 devkit with USB and FORTH processor that translates UART commands to bit-bang protocol, i.e. SPI. It works really fast but it's written in FORTH, so probably not really useful for you :)

**Comment by a-shafir** [comment]
@gkasprow i see io_reset commented in the code. probably some other changes.
Please also attach the pin assignment for the code.
So what was generating the DDS SPI command? 9910 devkit? I have no devkit.
Can you generate a dump (sequence) of the commands? 
Also what about the reading from DDS when all in parallel?

**Comment by a-shafir** [comment]
@gkasprow i see why in_sel not working to - it commented out
Good that i got this data from you )

**Comment by gkasprow** [comment]
The original DDS was working in parallel. So I didn't read the status from externally connected ones.
You can compare this code with one @jordens posted. I did not touch the pins assignment.
My CPLD project is [here](https://www.dropbox.com/sh/qj16dhavo4qxsw9/AAB5paEl0olZpSaRMjhjxSyOa?dl=0)

**Comment by gkasprow** [comment]
This is how typical setup wit hdevkit looks like. I this case I connected Sampler module, but the idea is same.
![20170915_102313](https://user-images.githubusercontent.com/4325054/33859026-586b3076-ded2-11e7-92f1-e032d6cf42ca.jpg)

**Comment by gkasprow** [comment]
this is the FORTH code I use to talk to LV595 registers.
The FORTH ascii is executed in FPGA and it toggling IOs routed to IO register with a few MHz rate. CPU runs with 50MHz clock.
The function below writes 16 bits via SPI to the register.


: spi_LV595 ( a ) 
	
	1 OUT0_SET_REG io! \ SCK hi
	1 16 lshift OUT0_SET_REG io! \ OE buff active
	1 2 lshift OUT0_CLR_REG io! \ RCLK low
	
    16 0 do
        dup
        1 15 lshift and
        if
           1 1 lshift OUT0_SET_REG io!
        else
           1 1 lshift OUT0_CLR_REG io!
        then
		1 OUT0_CLR_REG io! \ SCK low
		2*
		INP0_REG io@ 4 rshift 1 and +
        1 OUT0_SET_REG io! \ SCK hi	
    loop
	
	 1 2 lshift OUT0_SET_REG io! \ RCLK high
	 1 2 lshift OUT0_CLR_REG io! \ RCLK low
	 1 16 lshift OUT0_CLR_REG io! \ OE buff inactive
	 $ffffff and  \ return only valid bits
;

**Comment by gkasprow** [comment]
@a-shafir you are right - I had to modify the reset pin state to make it working.

**Comment by gkasprow** [comment]
And the code you have burned is in Urukul_CPLD_1ch.v file

**Comment by gkasprow** [comment]
@jordens here is your 9912 :)
![urukul9912](https://user-images.githubusercontent.com/4325054/33897155-fdf41d90-df64-11e7-9ba7-38506d371078.jpg)

**Comment by jordens** [comment]
@gkasprow Nice! Other than the missing amplifier, are there any other issues? Is the solder on the underside bad as well?

**Comment by hartytp** [comment]
Nice!

**Comment by gkasprow** [comment]
I got this amp separately. The PN is wrong so they got ones with straight leads and had to replace them.
It seems that the ICs they got had problems with wetting because other chips are fine. Anyway, it's their responsibility to solder it right

**Comment by jordens** [comment]
Are you returning the AD9912 boards to Technosystem to be fixed as well?

**Comment by gkasprow** [comment]
I will fix one piece myself to test it. Rest boards will be returned. I talked with the company and they say that it is quite standard thing that some chips have poor wetting and side contacts are not always covered by the tin. The flux that is part of the soldering paste is only roughly 10..15% of it and is not always able to clean the contact side. The joint still maintains the contact because it is low side of it which does the job. Side area does not need to be covered, but if we want, they can fix it.

**Comment by jordens** [comment]
IMHO it is likely that the joints are bad on the underside as well if they fail to wet that dramatically on the edge. And since they probably didn't inspect those joints there is no telling. I'd like to see them fix it.

**Comment by gkasprow** [comment]
I just desoldered the chip and the bottom pads are correctly soldered. Anyway it just looks non professional and should be fixed.

**Comment by a-shafir** [comment]
This is a long know for the technologists that too much solder is not helping and even can cover some problem. But in this case it is not about the professionalism but about Optical inspection standards. All the packages designed that way that there is possibility to ensure by inspection the proper electrically and mechanically soldering. For this package there must be wetting on the side edges of the pads. This indicates that there is wetting under the pads and also improves mechanical properties. 

For some of the packages we have (BGA and some of the qfn with signal pads under) the OIS will not enough. I think need to check if the factory can do x-ray inspection, they must have it in or out house at least for BGA but also for some regulator chips etc. Visual inspection of the QFN etc is the must. It shall not increase the cost of the samples production more than few %

Any way investing our time to solving the technological issues is the worst case scenario.

**Comment by a-shafir** [comment]
Urukul 9910, 4th channel, 10.000MHz, variable attenuation test, no termination:
![ds1z_quickprint1](https://user-images.githubusercontent.com/15978368/33939680-54e28f4a-e00c-11e7-810f-407db41730fc.png)

**Comment by gkasprow** [comment]
@a-shafir look [here](https://github.com/m-labs/sinara/issues/377#issuecomment-351379129)

**Comment by a-shafir** [comment]
@gkasprow ok, this is a news for me too. There are 3 similar looking "QFN-style" packages:
- one have sides plated to be good for wetting (some crystals for instance and old stile QFN)
- some have a etch/step around to have the wetting visible for OIS
- the newest are basicly BGA-like i.e. all wetting is under the package but just no balls.

That minds for us:
- Even we don't have a BGA on a board we need to inspect the packages are they are good for optical inspection (OIS) and if any one of them is not need to add X-ray inspection into the production list.
- Need to chose the parts for the boards more carreful. If the board normally not needs X-ray try to avoid any packages that have no OIS compatible pads (like LTM4622 etc) 
- It is good to put this as a part of the h/w and production specs checklist.

**Comment by a-shafir** [comment]
Gateware, core drivers and experiment for Urukul AD9910 rev 1.0 board.

The current setup is for KC705, see the details here:
https://github.com/m-labs/artiq/commit/dd3c8a336dc12cda33cadebdf3bafc77c0bc1565

The experiment initializes FMC DIO, the Urukul board and configures the single tone for all 4 channels.
Attenuator and SW are supported.

There are 2 "interference generator loops" (for DDS3 SPI and for ATT SPI). Each can be activated in the experiment. Note: with the current RTIO gateware it is not possible to generate the NU-mode intensive SPI communication so if any interference from the SPI clock will be seen on the output it will be not the worst case scenario.

The 4-channels DDS output:
![ds1z_quickprint11](https://user-images.githubusercontent.com/15978368/34085517-5dfc80e8-e392-11e7-8141-deb62ec9b3b2.png)

The setup:
![img_20171218_011736](https://user-images.githubusercontent.com/15978368/34085505-3a3653be-e392-11e7-982d-8c39d7c01d80.jpg)

**Comment by jordens** [comment]
* RF Switch rise time is roughly 100ns to 95%
* Spectrum looks ok. A few weird close in Spurs (+- 1.2 kHz and 11kHz at -70dBc) potentially related to grounding to be debugged. They are big when the board is on the antistatic bag on an optical table and otherwise disappear.

More data and screenshots to follow.

**Comment by dhslichter** [comment]
@a-shafir why are you using scope probes to look at the DDS outputs instead of SMA cables?

**Comment by gkasprow** [comment]
@jordens in such setup you easily receive and induce current from every short wave radio station including CB :)

**Comment by jordens** [comment]
@gkasprow in general yes, but not if you are in a room that is as insanely shielded as the one we are doing the tests in ;)

**Comment by a-shafir** [comment]
@dhslichter there was only one more SMA cable in the lab and no 50R terminators at all. Any way the probes are OK to check how the driver controls.

**Comment by jordens** [comment]
The screenshots are crappy and there is some analysis still to be done. I also have to parse my notes.
tl;dr: All good. The only issue are phase noise spurs at -75dBc at ~1.2 kHz offset with many harmonics, especially x11. We didn't figure out where exactly they came from. The spur offset is surprisingly low. Their offset is temperautre dependent (increases with temperature). And coupling different parts of the PCB (by finger), especially around the underside near the clock fanout, makes them worse. That spur needs to be debugged.

##  Attenuator step from 20 to 60 digital (32, 8, switch glitch)
![att_step_20-60](https://user-images.githubusercontent.com/1338946/34174075-fa708314-e4f7-11e7-9555-fdaa198e7c3a.png)
![att_step_20-60_b](https://user-images.githubusercontent.com/1338946/34174076-fab09ecc-e4f7-11e7-96be-9dd4d7ba0efb.png)

## Attenuator step from 31 to 32 digital (major carry glitch)
![att_step_31-32](https://user-images.githubusercontent.com/1338946/34174077-faf0f42c-e4f7-11e7-89ce-ba0e8667d573.png)

## Harmonics DC to 1 GHz, clock and ref leakage below RBW
![harmonics_0-1g_10dbm](https://user-images.githubusercontent.com/1338946/34174078-fb1c0374-e4f7-11e7-899a-19d71298ca74.png)

## Harmonics 2 and 3 vs power
![harmonics_6dbm](https://user-images.githubusercontent.com/1338946/34174079-fb4986e6-e4f7-11e7-921c-3a9df0837348.png)
![harmonics_10dbm](https://user-images.githubusercontent.com/1338946/34174080-fb7765d4-e4f7-11e7-9433-ea916f8d6e8d.png)

## Close in spurs (1.2 kHz multiples, especially 11x)
![narrow_sfdr_50k](https://user-images.githubusercontent.com/1338946/34174081-fba0057a-e4f7-11e7-85a3-a4a58704a287.png)

## PLL bump
![pll_bump_5m](https://user-images.githubusercontent.com/1338946/34174088-fd3ddbaa-e4f7-11e7-9909-cf176e59ef93.png)

## Phase noise 100 mHz to 10MHz
![phase_noise_100m-10m](https://user-images.githubusercontent.com/1338946/34174087-fcdb3c20-e4f7-11e7-8bf6-0f0a8540f17e.png)

## Phase noise 10 Hz to 10 MHz
![phase_noise_10-10m](https://user-images.githubusercontent.com/1338946/34174086-fc9c6ab8-e4f7-11e7-9a0d-110229eb8d67.png)

## Phase noise 1 kHz to 1MHz
![phase_noise_1k-1m](https://user-images.githubusercontent.com/1338946/34174085-fc5bef92-e4f7-11e7-8780-38d1784c7f30.png)

## RF switch turn on transient
![sw_on](https://user-images.githubusercontent.com/1338946/34174089-fdbd90ca-e4f7-11e7-9e77-e2a90cef2842.jpg)

## Nyquist rejection 400 MHz to 600 MHz
![nyquist_rejection_400m-600m](https://user-images.githubusercontent.com/1338946/34174082-fbc9c662-e4f7-11e7-80f5-81d7c7934f15.png)

## Nyquist rejection 450 MHz to 550 MHz
![nyquist_rejection_450m-550m](https://user-images.githubusercontent.com/1338946/34174083-fbf5c4c4-e4f7-11e7-861f-24ca7177d2ae.png)

**Comment by gkasprow** [comment]
@jordens such spur may be caused by PSU instability. Then 1.2kHz ripples caused by SMPS or LDO feedback loop may cause such spurs.
So you can have a look with scope on critical power rails.

**Comment by jordens** [comment]
Ack. We didn't find anything noticable on the rails. We'll do more prodding and poking.

**Comment by gkasprow** [comment]
@jordens I can have a look with spectrum analyzer or when necessary with SSA. Is ADI devkit sufficient to recreate this spur?

**Comment by jordens** [comment]
I don't know. But I'd be surprised if you see them on the devkit.
I'll give you details about the setup later.

**Comment by gkasprow** [comment]
No, I mean Urukul controlled by ADI devkit instead of the Artiq

**Comment by jordens** [comment]
I think you should see them on Urukul with yout setup. I added the description of the setup to the post above.

**Comment by a-shafir** [comment]
@jordens what reference clock used for the test? KC705 output from the FPGA pll possible has the low freq modulation from the power or even fan. And the AD9910 pll most likely will unable to clean it.

**Comment by jordens** [comment]
@a-shafir the reference is clean. And you hypothesis is already excluded be the other observations.

**Comment by jordens** [comment]
AD9912 variant also works fine.
@gkasprow did you get a chance to hunt for the 1.2 kHz spurs?

**Comment by gkasprow** [comment]
@jordens not yet. The University and my lab was closed during Xmass. I could not even cross the main gate. They open tomorrow.

**Comment by jordens** [comment]
Ack. That's fine!

**Comment by jordens** [comment]
The spur might be a beat due to the on board oscillator leaking through the mux (-61 dB isolation, 25 ppm spec, 1.2 kHz is 12 ppm). But the magnitude is still high: the capacitor population should have created much more isolation. The symptoms are otherwise fitting. @gkasprow if you get to testing this before I do, try removing L9.

**Comment by hartytp** [comment]
@jordens Based on those measurements are you happy to scrap the screening clips for the next revision? We could also DNP them, but scrapping them frees up quite a bit of space (and, anyway, AFAICT, there wasn't enough space on Urukul to do the screening properly anyway, so the current arrangement is a bit of a bodge!). Or, do you want to take measurements with the screening in place first to see if it actually makes a difference?

**Comment by jordens** [comment]
Why would you remove the clips now? The pending changes certainly don't need the additional space. And yes, I'd like to test and quantify the crosstalk in a rack with Urukul in proximity to other Urukuls, Kaslis, or DIOs pulsing 5 V into 50 Ohm.

**Comment by hartytp** [comment]
> Why would you remove the clips now?

No hurry for this, just came up now because I was reviewing my to do list for these boards.

My guess is that most of the screening we've added to the EEMs is unnecessary and can be removed at some point once we've finished testing, but there is no rush for that.

**Comment by gkasprow** [comment]
@jordens tomorrow I plan to spend full day  on this issue.

**Comment by gkasprow** [comment]
@jordens I got such spectra on my setup
![tek00000](https://user-images.githubusercontent.com/4325054/34774850-4dae6152-f611-11e7-826f-de191697eeda.png)

**Comment by jordens** [comment]
@gkasprow those are not the ones I am interested in. Look above for a description, photos and analysis of the ~kHz spurs.

**Comment by gkasprow** [comment]
above there are mainly setup windows :)
![tek00002](https://user-images.githubusercontent.com/4325054/34775549-dd6b245e-f613-11e7-889d-5d3d848c44a5.png)

**Comment by gkasprow** [comment]
I  brought a spectrum analyser (N9030A) with SSA function and got such plot. I confirm that the level of spurs gets increased by 20dB when I touch the clock chip

![screen_0001](https://user-images.githubusercontent.com/4325054/34778714-6f64e886-f61e-11e7-9fc1-0ee02de31e98.png)

**Comment by gkasprow** [comment]
I removed L9 and it helped. So in next revision will add MOSFET switch.
![Uploading Screen_0002.png…]()

**Comment by gkasprow** [comment]
![screen_0002](https://user-images.githubusercontent.com/4325054/34779330-29cc34d0-f620-11e7-9e35-419d30e3961d.png)

**Comment by jordens** [comment]
Excellent, Greg. Yes. A mosfet would work.

**Comment by gkasprow** [comment]
@jordens do you want to control MLVDS receiver type via FPGA pin ?

**Comment by jordens** [comment]
Yes. If there is a pin available.

**Comment by hartytp** [comment]
Closing since Urukul v1.1 has gone to manufacture, so I assume Urkul v1.0 testing is complete (reopen if I'm wrong about that).

**Comment by jordens** [comment]
I still want those tests.

**Comment by hartytp** [comment]
Okay, sorry. Overly enthusiastic in my pruning.

**Comment by jordens** [comment]
No problem.

**Comment by jordens** [comment]
Another quick shot at verifying that the clock distribution with Kasli and Urukul fundamentally works (modulo the known issues). Instrument noise floor still at -130 dBc/Hz. I would expect flicker phase (0.9 GS/s, 24x PLL, 150 MHz carrier, L(1kHz)=-116 dBc/Hz) to be a bit better given the datasheet graph (1 GS/s, 20x PLL, 200 MHz carrier, L(1kHz)=-120 dBc/Hz).

![image](https://user-images.githubusercontent.com/1338946/35699908-b18afd0c-0791-11e8-80c0-e9f200f0498d.png)

**Comment by hartytp** [comment]
We took some data today using the AD9910 Urukul with a Wenzel oscillator source. Will upload tomorrow.

**Comment by hartytp** [comment]
Courtesy of @WeiDaZhang: Urukul AD9910 phase noise.
[Urukul Phase Noise Measurement - Ver1.0 w- XTAL Disabled.pdf](https://github.com/m-labs/sinara/files/1689973/Urukul.Phase.Noise.Measurement.-.Ver1.0.w-.XTAL.Disabled.pdf)

**Comment by hartytp** [comment]
Urukul noise performance basically matches the data sheet curve.

**Comment by hartytp** [comment]
[Urukul Amplitude Modulation Noise Measurement - Ver1.0 w- XTAL Disabled.pdf](https://github.com/m-labs/sinara/files/1690366/Urukul.Amplitude.Modulation.Noise.Measurement.-.Ver1.0.w-.XTAL.Disabled.pdf)

Uruku AM noise. [cf](https://martein.home.xs4all.nl/pa3ake/hmode/dds_ad9910_amnoise.html)

**Comment by hartytp** [comment]
Conclusions:
- the loop filter isn't quite optimal, which isn't a surprise since the VCO characteristics aren't particularly well specified (not even clear how well controlled they are). We could probably make the servo bump about 7dB smaller with optimisation, but probably not worth the effort.
- the LDO dropout voltage doesn't seem to have too much of an impact. There are a few spurs on the AM output, which may be SMPS leakage. Might be worth adding inductors between the SMPS and LDOs (cheap to do).

Other than that, all looks really good! Congratulations @gkasprow

**Comment by hartytp** [comment]
NB haven't bothered to check performance without the PLL. We may not quite hit the noise floor of the DDS in that case, but clearly there is no drastic problem.

**Comment by jbqubit** [comment]
Why did @WeiDaZhang switch to higher phase noise [E4422B](https://www.keysight.com/en/pd-1000002806%3Aepsg%3Apro-pn-E4422B/esg-a-series-analog-rf-signal-generator-4-ghz?cc=US&lc=eng) Ref vs Wenzel?

Looks like <5 dB deviation from Datasheet at single frequency. Nice job! Is it similar at other output frequencies?

**Comment by hartytp** [comment]
Before I forget: @cjbe tested phase synchronisation with FPGA. All looks good!

**Comment by jordens** [comment]
Nice. @cjbe @hartytp with the FPGA driving SYNC? Or just among the DDS?

**Comment by cjbe** [comment]
@jordens with FPGA driving sync. I tested synchronisation between channels on the same board and channels on different boards. 

The DDS sync signal was generated by a TTLClkGen, and IO_UPDATE by a TTL SERDES output. I adjusted the sync input delay on each channel to sit in the centre of the eye - thus the SYNC_CLK of each DDS is aligned to better than 1 DDS clk period (=1ns @ 1GHz).

I choose an offset between IO_UPDATE and the sync TTLClkGen such that IO_UPDATE aligned roughly on the falling edge of the DDS SYNC_CLK. (I did not look at the CPLD timing report to work out if this would work with a fixed value over PVT).

I saw no synchronisation errors over ~10^10 IO_UPDATES, but I did see at least one SYNC_SMP_ERR from most of the AD9910 over this 10^10 events using the standard 300ps validation window - I have not looked to see if this happens with slightly smaller validation windows.

**Comment by jordens** [comment]
Ok. Nice!

**Comment by hartytp** [comment]
> (I did not look at the CPLD timing report to work out if this would work with a fixed value over PVT).

This still concerns me, and is something we need to look at. @jordens do you know how well constrained this timing is atm?

**Comment by jordens** [comment]
SYNC is not routed through the CPLD. That can't be related to SYNC_SMP_ERRs. How are you checking for IO_UPDATE synchronization errors? I don't know about the constraints there (other than the 7.7ns nominal propagation delay) and whether that's even specified.

**Comment by cjbe** [comment]
There are two considerations here:

1) SYNC_SMP_ERRs:
This is just due to the jitter of the FPGA generated SYNC relative to the AD9910 1GHz clock. I see a narrower eye using the FPGA generated SYNC vs the AD9910 sync (~6 taps = ~450ps for the FPGA generated clock). I am using a validation window of 300ps, so it is not surprising that at least 1 in 10^10 events closes the eye to <300ps, but may be fine for a validation window one tap smaller.

2) IO_UPDATE sync errors:
I am setting CFR1 bit 11 so the phase resets on IO_UPDATE, I am then triggering a scope on a signal synchronous with IO_UPDATE, and looking at the RF phase relative to this trigger with a scope on persist.    As the RTIO fine clock is 1ns, and the SYNC_CLK period is 4ns, there are 4 possible IO_UPDATE alignments. With one of these I see frequent phase alignment errors on the scope, but with the other 3 I have not seen any - I choose the best out of these three by looking at the IO_UPDATE timing at the AD9910. 

My only worry is how large the PVT change in the 7.7ns propagation delay is - if it is larger than ~2ns (which is probably is) things are getting marginal (the IO_UPDATE setup/hold is 1.75ns/0ns).
We could register IO_UPDATE on the DDS0_SYNC_CLK at the CPLD - from the CPLD datasheet it looks like the PVT on the clock input buffer and standard output buffer sum to ~0.9ns, so this may well work out.

**Comment by jordens** [comment]
That's what I thought. I don't think P is relevant for us (you'll have to tweak it anyway). And I don't think VT are more than 500 ps for our VT variability.
Registering io update doesn't help AFAICT because (1) there is no way to delay the cpld output to meet sync clk s/h and (2) the s/h window on the cpld is unlikely to be significantly larger than the 2.25 ns of the DDS.

**Comment by hartytp** [comment]
NB we have an IO_UPDATE return on EEM1 in the current design, so if we see any PVT issues with the phase synchronisation, we should be able to use that to calibrate the delays in the CPLD buffers. Given that, I no longer have any concerns about phase synchronisation with Urukul!

**Comment by hartytp** [comment]
What do people think about maybe sprinkling a few cheap, low-profile SMT/THT heatsinks on Urukul (and maybe/other EEMs with significant power consumption) near the most power-hungry ICS?

**Comment by jordens** [comment]
I think it's ok. I'm just surprised the fanout gets so hot compared to the DDS.

**Comment by gkasprow** [comment]
THere is a place under on the other side for small ceramic heatsink

**Comment by marmeladapk** [comment]
_Migrated to sinara-hw/Urukul#3_
